src/dex/nerve/nerve.ts

```diff
@@ -0,0 +1,405 @@
+        : blockNumber;
+
+    return Promise.all(
+      _pools.map(async eventPool => {
+        let state = eventPool.getState(_blockNumber);
+        if (!state || !state.isValid) {
+          this.logger.info(
+            `State for ${this.dexKey} pool ${eventPool.name} is stale or invalid on block ${_blockNumber}. Generating new one`,
+          );
+          const newState = await eventPool.generateState(_blockNumber);
+          eventPool.setState(newState, _blockNumber);
+          return { state: newState, pool: eventPool };
+        } else {
+          return { state, pool: eventPool };
+        }
+      }),
+    );
+  }
+
+  async getPoolIdentifiers(
+    srcToken: Token,
+    destToken: Token,
+    side: SwapSide,
+    blockNumber: number,
+  ): Promise<string[]> {
+    return this.allPools
+      .filter(pool => {
+        return (
+          pool.tokenAddresses.includes(srcToken.address.toLowerCase()) &&
+          pool.tokenAddresses.includes(destToken.address.toLowerCase())
+        );
+      })
+      .map(pool => Nerve.getIdentifier(this.dexKey, pool.address));
+  }
+
+  async getPricesVolume(
+    srcToken: Token,
+    destToken: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    blockNumber: number,
+    limitPools?: string[],
+  ): Promise<null | ExchangePrices<NerveData>> {
+    try {
+      if (side === SwapSide.BUY) return null;
+
+      const _srcToken = wrapETH(srcToken, this.network);
+      const _destToken = wrapETH(destToken, this.network);
+
+      if (
+        _srcToken.address.toLowerCase() === _destToken.address.toLowerCase()
+      ) {
+        return null;
+      }
+
+      const filterPoolsByIdentifiers = (
+        identifiers: string[],
+        pools: EventPoolOrMetapool[],
+      ) => {
+        return pools.filter(pool =>
+          identifiers.includes(Nerve.getIdentifier(this.dexKey, pool.address)),
+        );
+      };
+
+      const selectedPools = !limitPools
+        ? filterPoolsByIdentifiers(
+            await this.getPoolIdentifiers(
+              _srcToken,
+              _destToken,
+              side,
+              blockNumber,
+            ),
+            this.allPools,
+          )
+        : filterPoolsByIdentifiers(limitPools, this.allPools);
+
+      const statePoolPair = await this.getStates(selectedPools, blockNumber);
+
+      // here side === SwapSide.SELL
+      const unitVolume = getBigIntPow(10 ** _srcToken.decimals);
+      const chunks = amounts.length - 1;
+
+      const _width = Math.floor(chunks / NERVE_CHUNKS);
+      const _amounts = [unitVolume].concat(
+        Array.from(Array(NERVE_CHUNKS).keys()).map(
+          i => amounts[(i + 1) * _width],
+        ),
+      );
+
+      const result: ExchangePrices<NerveData> = [];
+      for (const { pool, state } of statePoolPair) {
+        // To be able to update state property isValid if calculations receive error
+        const _state = _.cloneDeep(state) as PoolState;
```

Comment:
I don't think you need to clone the state! You simply set the invalid flag to true but you don't reuse the old state. I suggested a type change below for poolState

Reply:
Right now, the NervePoolMath functions are allowed to modify the state, so cloning it makes sense, but I would suggest making those functions take a DeepReadonly<State> instead and not modify the state, only throw the exception, so that there is no need to clone, then the exception handlers can set the isValid flag to false (in pool event handlers, I think this is already done).

Reply:
Fixed another way as we discussed